Essentials/GridFinder.cs

```diff
@@ -0,0 +1,321 @@
+        }
+
+        private static bool IsPlayerIdCorrect(long playerId, List<MyCubeGrid> gridList) {
+
+            MyCubeGrid biggestGrid = null;
+
+            foreach (var grid in gridList)
+                if (biggestGrid == null || biggestGrid.BlocksCount < grid.BlocksCount)
+                    biggestGrid = grid;
+
+            /* No biggest grid should not be possible, unless the gridgroup only had projections -.- just skip it. */
+            if (biggestGrid == null)
+                return false;
+
+            bool hasOwners = biggestGrid.BigOwners.Count != 0;
+
+            if(!hasOwners) {
+
+                if (playerId != 0L)
+                    return false;
+
+                return true;
+            }
+
+            return playerId == biggestGrid.BigOwners[0];
+        }
+
+        public static List<MyCubeGrid> FindGridList(string gridNameOrEntityId, MyCharacter character, bool includeConnectedGrids) {
+
+            List<MyCubeGrid> grids = new List<MyCubeGrid>();
+
+            if (gridNameOrEntityId == null && character == null)
+                return new List<MyCubeGrid>();
+
+            if (includeConnectedGrids) {
+
+                ConcurrentBag<MyGroups<MyCubeGrid, MyGridPhysicalGroupData>.Group> groups;
+
+                if (gridNameOrEntityId == null)
+                    groups = FindLookAtGridGroup(character);
+                else
+                    groups = FindGridGroup(gridNameOrEntityId);
+
+                if (groups.Count > 1)
+                    return null;
+
+                foreach (var group in groups) {
+                    foreach (var node in group.Nodes) {
+
+                        MyCubeGrid grid = node.NodeData;
+
+                        if (grid.Physics == null)
+                            continue;
+
+                        grids.Add(grid);
+                    }
+                }
+
+            } else {
+
+                ConcurrentBag<MyGroups<MyCubeGrid, MyGridMechanicalGroupData>.Group> groups;
+
+                if (gridNameOrEntityId == null)
+                    groups = FindLookAtGridGroupMechanical(character);
+                else
+                    groups = FindGridGroupMechanical(gridNameOrEntityId);
+
+                if (groups.Count > 1)
+                    return null;
+
+                foreach (var group in groups) {
+                    foreach (var node in group.Nodes) {
+
+                        MyCubeGrid grid = node.NodeData;
+
+                        if (grid.Physics == null)
+                            continue;
+
+                        grids.Add(grid);
+                    }
+                }
+            }
+
+            return grids;
+        }
+
+        public static ConcurrentBag<MyGroups<MyCubeGrid, MyGridPhysicalGroupData>.Group> FindGridGroup(string gridName) {
+
+            ConcurrentBag<MyGroups<MyCubeGrid, MyGridPhysicalGroupData>.Group> groups = new ConcurrentBag<MyGroups<MyCubeGrid, MyGridPhysicalGroupData>.Group>();
```

Comment:
Use var for declarations where the type is obvious from the assignment so lines don't get too long